### PR TITLE
frontend/fix: #213 added emeregency contacts feature and integrated apis

### PR DESCRIPTION
**Pull request by chakri642**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: Frontend/ui-customer/src/Components/ContactList/View.purs [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+                                    , margin (Margin 0 23 16 23)
+                                    , padding (Padding 0 0 0 0)
+                                    , gravity CENTER
+                                    ]
+                                    [ imageView
+                                        [ if item.isSelected then height $ V 24 else height $ V 17
+                                        , if item.isSelected then width $ V 24 else width $ V 17
+                                        , if item.isSelected then imageWithFallback "ny_ic_selected_icon,https://assets.juspay.in/nammayatri/images/user/ny_ic_selected_icon" else imageWithFallback "ny_ic_outer_circle,https://assets.juspay.in/nammayatri/images/user/ny_ic_outer_circle"
+                                        , margin (Margin 0 0 0 0)
+                                        , padding (Padding 0 0 0 0)
```
Comment (ashishsingh101) [inline Frontend/ui-customer/src/Components/ContactList/View.purs:254]: remove this

File: Frontend/ui-customer/src/Components/ContactList/View.purs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+          }
+        , suffixImageConfig
+          { visibility = GONE
+          }
+        }
+  in
+    genericHeaderConfig'
+
+horizontalLine :: Int -> Int -> forall w. PrestoDOM (Effect Unit) w
+horizontalLine marginLeft marginRight =
```
Comment (ashishsingh101) [inline Frontend/ui-customer/src/Components/ContactList/View.purs:139]: remove `marginLeft` and `marginRight`

File: Frontend/ui-customer/src/Components/ContactList/View.purs [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+              , width MATCH_PARENT
+              , orientation VERTICAL
+              ]
+              [ linearLayout
+                  [ height WRAP_CONTENT
+                  , width MATCH_PARENT
+                  , orientation HORIZONTAL
+                  , padding (Padding 16 12 0 12)
+                  , background Color.blue600
+                  , visibility if (((filter (\contact -> DS.take 1 contact.name == letter) config.contactsData) /= []) && (config.editedText == "")) then VISIBLE else GONE
```
Comment (ashishsingh101) [inline Frontend/ui-customer/src/Components/ContactList/View.purs:181]: if you will use `contains` then modify this as well

File: Frontend/ui-customer/src/Helpers/Utils.purs [hunk 4]
```diff
@@ -84,6 +84,8 @@ foreign import storeCallBackCustomer :: forall action. (action -> Effect Unit) -
 foreign import getLocationName :: forall action. (action -> Effect Unit) -> String -> String -> String -> (String -> String -> String -> action) -> Effect Unit
 
 foreign import getCurrentDate :: String -> String
+foreign import storeCallBackContacts :: forall action. (action -> Effect Unit) -> ((Array Contacts) -> action) -> Effect Unit
+foreign import stringToJsonArray :: String -> (Array NewContacts)
```
Comment (ashishsingh101) [inline Frontend/ui-customer/src/Helpers/Utils.purs:88]: can you rename it to some other name like `parseNewContacts`

**Review by ashishsingh101 — CHANGES_REQUESTED** [review]

File: Frontend/ui-customer/src/Components/ContactList/View.purs (hunk 3 above)
Comment (chakri642) [reply Frontend/ui-customer/src/Components/ContactList/View.purs:181]: Here we need only first letter to check so i am using filter.

File: Frontend/ui-customer/src/Screens/CustomerUtils/EmergencyContactsScreen/View.purs [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+        , width $ WRAP_CONTENT
+        , text "Remove"
+        , color Color.blue900
+        , textSize 14
+        , onClick push (const (RemoveButtonClicked contact))
+        ]
+    ]
+
+--------------------------------------------------- genericHeaderConfig -----------------------------------------------------
+genericHeaderConfig :: EmergencyContactsScreenState -> GenericHeader.Config
```
Comment (ashishsingh101) [inline Frontend/ui-customer/src/Screens/CustomerUtils/EmergencyContactsScreen/View.purs:235]: can we add configs in new file

File: Frontend/ui-customer/src/Screens/CustomerUtils/EmergencyContactsScreen/View.purs [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+                , gravity CENTER
+                , padding (Padding 16 0 16 0)
+                -- , visibility if state.props.showContactList then GONE else VISIBLE
+                ]
+                [ emergencyContactsView push state
+                , PrimaryButton.view (push <<< PrimaryButtonActionControll) (primaryButtonConfig state)
+                ]
+            ]
+        , if state.props.showContactList then (contactListView push state) else emptyTextView state
+        , if (state.props.showInfoPopUp == true) then removeContactPopUpView push state else textView []
```
Comment (ashishsingh101) [inline Frontend/ui-customer/src/Screens/CustomerUtils/EmergencyContactsScreen/View.purs:88]: use emptyTextView here

File: Frontend/ui-customer/src/Components/SettingSideBar/View.purs [hunk 7]
```diff
@@ -172,14 +176,15 @@ settingsMenuView item push  =
                               SETTINGS_LANGUAGE       -> ChangeLanguage 
                               SETTINGS_ABOUT          -> GoToAbout 
                               SETTINGS_LOGOUT         -> OnLogout
-                              SETTINGS_SHARE_APP      -> ShareAppLink)
+                              SETTINGS_SHARE_APP      -> ShareAppLink
```
Comment (ashishsingh101) [inline Frontend/ui-customer/src/Components/SettingSideBar/View.purs:179]: remove this

**Review by ashishsingh101 — CHANGES_REQUESTED** [review]

File: Frontend/ui-customer/src/Screens/CustomerUtils/EmergencyContactsScreen/Controller.purs [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+          _ <- launchAff_ $ flowRunner $ do 
+              _ <- toggleLoader false
+              pure unit
+          pure NoAction
+        ]
+    else do
+      let newContacts = getContactList contacts
+      localContacts <- pure $ getValueToLocalStore CONTACTS
+      contactsInJson <- pure $ parseNewContacts localContacts
+      let list1 = filter (\x -> filter (\y -> x.number<>x.name == y.number<>y.name) contactsInJson == [] ) newContacts
```
[the remark this answers is not in the document]
Comment (chakri642) [reply Frontend/ui-customer/src/Screens/CustomerUtils/EmergencyContactsScreen/Controller.purs:104]: i written that list1 is to remove the 3 contacts from the list of contacts

**Comment by witcher-shailesh** [comment]
@chakri642 can you squash the commits as well?

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/CommonJsInterface.java [hunk 9]
```diff
@@ -3596,4 +3600,36 @@ private String getAPIResponse(String url) {
...
+                    contactsStoreCall(MainActivity.getInstance().getJuspayServices().getDuiCallback(), contacts);
+                }
+            }
+        } catch (Exception e) {
+            Log.e("error inside contactPermission", String.valueOf(e));
+        }
+    }
+
+
+    static String storeCallBContact = "dummystoreCallBackContacts";
```
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/CommonJsInterface.java:3621]: can we make it null by default?

**Comment by ks-shree** [comment]
> ## Type of Change
> 
>     * [ ]  Bugfix
> 
>     * [x]  New feature
> 
>     * [ ]  Enhancement
> 
>     * [ ]  Refactoring
> 
>     * [ ]  Dependency updates
> 
> 
> ## Description
> ### Additional Changes
> 
>     * [ ]  This PR modifies the database schema (database migration added)
> 
>     * [ ]  This PR modifies dhall configs/environment variables
> 
>     * [ ]  This PR contains API breaking changes
> 
> 
> ## Motivation and Context
> ## How did you test it?
> ## Checklist
> 
>     * [ ]  I formatted the code and addressed linter errors `./dev/format-all-files.sh`
> 
>     * [ ]  I reviewed submitted code
> 
>     * [ ]  I added unit tests for my changes where possible
> 
>     * [ ]  I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

File: Frontend/ui-customer/src/Screens/CustomerUtils/EmergencyContactsScreen/Controller.purs [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    AfterRender -> trackAppScreenRender appId "screen" (getScreen EMERGENCY_CONTACS_SCREEN)
+    BackPressed -> do
+      trackAppBackPress appId (getScreen EMERGENCY_CONTACS_SCREEN)
+      trackAppEndScreen appId (getScreen EMERGENCY_CONTACS_SCREEN)
+    GenericHeaderActionController act -> case act of
+      GenericHeader.PrefixImgOnClick -> do
+        trackAppActionClick appId (getScreen EMERGENCY_CONTACS_SCREEN) "generic_header_action" "back_icon_onclick"
+        trackAppEndScreen appId (getScreen EMERGENCY_CONTACS_SCREEN)
+      _ -> pure unit
+    _ -> pure unit
```
Comment (ks-shree) [inline Frontend/ui-customer/src/Screens/CustomerUtils/EmergencyContactsScreen/Controller.purs:52]: @chakri642 Remove default case and add logs for all other actions

**Review by witcher-shailesh — APPROVED** [review]

**Review by ashishsingh101 — APPROVED** [review]

**Pull request merged**